zebra-chain/src/serialization/compact_size.rs

```diff
@@ -0,0 +1,350 @@
+            Err(Parse("compactsize larger than protocol message limit"))?;
+        }
+
+        Ok(CompactSizeMessage(size))
+    }
+}
+
+impl From<CompactSizeMessage> for u64 {
+    #[inline]
+    fn from(size: CompactSizeMessage) -> Self {
+        size.0.into()
+    }
+}
+
+impl From<CompactSize64> for u64 {
+    #[inline]
+    fn from(size: CompactSize64) -> Self {
+        size.0
+    }
+}
+
+impl From<u64> for CompactSize64 {
+    #[inline]
+    fn from(size: u64) -> Self {
+        CompactSize64(size)
+    }
+}
+
+impl ZcashSerialize for CompactSizeMessage {
+    /// Serialize a CompactSizeMessage into the Zcash consensus-critical format.
+    ///
+    /// # Panics
+    ///
+    /// If the value exceeds `MAX_PROTOCOL_MESSAGE_LEN`.
+    #[inline]
+    fn zcash_serialize<W: std::io::Write>(&self, writer: W) -> Result<(), std::io::Error> {
+        // # Security
+        // Defence-in-depth for memory DoS via preallocation.
+        if self.0
+            > MAX_PROTOCOL_MESSAGE_LEN
+                .try_into()
+                .expect("usize fits in u64")
+        {
+            // This is validated data inside Zebra, so we panic.
+            panic!("compactsize larger than protocol message limit");
+        }
+
+        // Use the same serialization format as CompactSize64.
+        let size: u64 = self.0.into();
+        CompactSize64(size).zcash_serialize(writer)
+    }
+}
+
+impl ZcashDeserialize for CompactSizeMessage {
+    #[inline]
+    fn zcash_deserialize<R: std::io::Read>(reader: R) -> Result<Self, SerializationError> {
+        // Use the same deserialization format as CompactSize64.
+        let size: CompactSize64 = reader.zcash_deserialize_into()?;
+
+        size.0.try_into()
+    }
+}
+
+impl ZcashSerialize for CompactSize64 {
+    #[inline]
+    fn zcash_serialize<W: std::io::Write>(&self, mut writer: W) -> Result<(), std::io::Error> {
+        let n = self.0;
+        match n {
+            0x0000_0000..=0x0000_00fc => writer.write_u8(n as u8),
+            0x0000_00fd..=0x0000_ffff => {
+                writer.write_u8(0xfd)?;
+                writer.write_u16::<LittleEndian>(n as u16)
+            }
+            0x0001_0000..=0xffff_ffff => {
+                writer.write_u8(0xfe)?;
+                writer.write_u32::<LittleEndian>(n as u32)
+            }
+            _ => {
+                writer.write_u8(0xff)?;
+                writer.write_u64::<LittleEndian>(n)
+            }
+        }
+    }
+}
+
+impl ZcashDeserialize for CompactSize64 {
+    #[inline]
+    fn zcash_deserialize<R: std::io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        use SerializationError::Parse;
+
+        let flag_byte = reader.read_u8()?;
+        let size = match flag_byte {
+            n @ 0x00..=0xfc => Ok(n as u64),
+            0xfd => match reader.read_u16::<LittleEndian>()? {
+                n @ 0x0000_00fd..=0x0000_ffff => Ok(n as u64),
```

Comment:
Nit: not sure if this makes sense, but I thought it could improve readability:

```suggestion
                n @ 0x00fd..=0xffff => Ok(n as u64),
```

Reply:
What do you think of 1fb3d2d70?

I tried to make the patterns consistent with the sizes of the integers they're matching.
I also used symbolic constants where possible. (It doesn't work unless the types match exactly.)

Reply:
Looks good!

Regarding types not matching, we could use things like `u32::from(u16::MAX + 1)`, but I don't think it adds much value. We could put these values in constants, but I'm not sure if it improves things.

We can leave it as is because it's already great, and maybe think about this in the future if we end up revisiting this :thinking: